### PR TITLE
Fix file move for custom editors

### PR DESCRIPTION
**Pull request by mjbvz**
**Problem**
The working copy for custom editors uses a special scheme for its resource: `vscode-custom-editor` . However this broke file move. I investigate and determined the root cause:

- The WorkingCopyService monitors file change events: https://github.com/microsoft/vscode/blob/17cc949013df9abb021da74c77b9508cf8607849/src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts#L193

- When one occurs, it tries to get all the dirty working copies for the moved resource

- However since our custom editor working copy now uses a `vscode-custom-editor` resource, we never match the moved resource.

**Fix**
My fix was to add an `editorResource` to the working copy. This is the actual resource on disk that the working copy represents.

I don't think this is a great fix. Please let me know if you have any other ideas on how to fix this @bpasero

### CODE REVIEW

**Comment by bpasero** [comment]
@mjbvz agree, this is not a good route. And it shows the actual issue: you want to rely on a generic file listener for moving non-text-based custom editors and I think we cannot do that. After all, we actually do not know what underlying resource the custom editor is operating on and if a move should be initiated based on file events.

Instead, I think this needs to be driven entirely from the extension by doing the listening on their end. Maybe the custom editors has 5 editors it needs to listen to? Or maybe it needs to listen on a DB connection for changes?

**Comment by mjbvz** [comment]
@bpasero I've updated the change based on our discussion this morning

File: src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts [hunk 1]
```diff
@@ -142,9 +145,10 @@ export interface IWorkingCopyFileService {
 
 	//#endregion
 
-
 	//#region Path related
 
+	registerWorkingCopyProvider(provider: WorkingCopyProvider): void;
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts]: This deserves a JSDoc comment what the intent is.

File: src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts [hunk 2]
```diff
@@ -275,7 +279,18 @@ export class WorkingCopyFileService extends Disposable implements IWorkingCopyFi
 
 	//#region Path related
 
+	private readonly workingCopyProviders: WorkingCopyProvider[] = [];
+
+	registerWorkingCopyProvider(provider: WorkingCopyProvider): void {
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts]: This should return a disposable to unregister.

File: src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts [hunk 3]
```diff
@@ -275,7 +279,18 @@ export class WorkingCopyFileService extends Disposable implements IWorkingCopyFi
...
+		this.workingCopyProviders.push(provider);
+	}
+
 	getDirty(resource: URI): IWorkingCopy[] {
+		const providerResources = flatten(this.workingCopyProviders.map(provider => provider(resource)));
+		if (providerResources.length) {
+			return providerResources;
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts]: Why does this return early? We need to ask the `IWorkingCopyService` as well, no?

**Review by bpasero — CHANGES_REQUESTED** [review]
@mjbvz some comments inline and here:
* please make sure to have this covered in tests from the working copy file service tests file
* the `IWorkingCopyFileService#getDirty()` method would probably be simpler if the working copy file service would simply register a provider itself from the constructor that would lookup the dirty working copies with the code as today from the working copy service? so you could argue it would be a default provider that is just always there and then on top there can be additional providers

**Comment by mjbvz** [comment]
Thanks for the feedback @bpasero. I've update the PR with your suggested changes

File: src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts [hunk 4]
```diff
@@ -142,9 +148,15 @@ export interface IWorkingCopyFileService {
 
 	//#endregion
 
-
 	//#region Path related
 
+	/**
+	 * Register a new provider for working copies.
+	 *
+	 * @return Disposable the unregisters the provider.
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts]:
```suggestion
	 * @return a disposable that unregisters the provider.
```

File: src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts [hunk 5]
```diff
@@ -142,9 +148,15 @@ export interface IWorkingCopyFileService {
 
 	//#endregion
 
-
 	//#region Path related
 
+	/**
+	 * Register a new provider for working copies.
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts]:
```suggestion
	 * Register a new provider for working copies based on a resource.
```

File: src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts [hunk 6]
```diff
@@ -180,6 +192,18 @@ export class WorkingCopyFileService extends Disposable implements IWorkingCopyFi
...
+				if (this.fileService.canHandleResource(resource)) {
+					// only check for parents if the resource can be handled
+					// by the file system where we then assume a folder like
+					// path structure
+					return isEqualOrParent(workingCopy.resource, resource);
+				}
+
+				return isEqual(workingCopy.resource, resource);
+			});
+		});
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/common/workingCopyFileService.ts]:
```suggestion

		// register a default working copy provider that uses the working copy service
		this.registerWorkingCopyProvider(resource => {
			return this.workingCopyService.workingCopies.filter(workingCopy => {
				if (this.fileService.canHandleResource(resource)) {
					// only check for parents if the resource can be handled
					// by the file system where we then assume a folder like
					// path structure
					return isEqualOrParent(workingCopy.resource, resource);
				}

				return isEqual(workingCopy.resource, resource);
			});
		});
```

File: src/vs/workbench/services/workingCopy/test/common/workingCopyService.test.ts [hunk 7]
```diff
@@ -12,64 +12,65 @@ import { TestWorkingCopyService } from 'vs/workbench/test/common/workbenchTestSe
...
+		this._onDispose.fire();
 
-			super.dispose();
-		}
+		super.dispose();
 	}
+}
+
+suite('WorkingCopyService', () => {
+
```
Comment (bpasero) [inline src/vs/workbench/services/workingCopy/test/common/workingCopyService.test.ts]:
```suggestion
suite('WorkingCopyService', () => {
```

**Review by bpasero — APPROVED** [review]
Looks good, please apply the style changes I made and we are good to go.

**Pull request merged**